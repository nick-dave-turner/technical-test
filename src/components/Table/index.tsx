import React from "react";
import styled from "styled-components";

const StyledTable = styled.table`
  width: 100%;
  border-spacing: 0;
  border-collapse: collapse;
  border-bottom: 2px solid #ddd;
`;

const Td = styled.td`
  padding: 0.625rem;
  border: 1px solid #ddd;
  background: white;
`;

const Th = styled(Td)`
  padding: 0.625rem;
  text-align: left;
  font-weight: bold;
  background: #ededed;
`;

const NoData = styled.p`
  margin: 0;
  text-align: center;
`;

type TableProps = {
  colDefs: Array<string>;
  rowData: Array<any>;
};

const Table: React.FC<TableProps> = ({ colDefs, rowData }) => {
  return (
    <StyledTable>
      <thead>
        <tr>
          {colDefs.map((col, index) => (
            <Th key={index}>{col}</Th>
          ))}
        </tr>
      </thead>
      <tbody>
        {rowData.length === 0 && (
          <tr>
            <Td colSpan={colDefs.length}>
              <NoData>You currently have no entries to display.</NoData>
            </Td>
          </tr>
        )}

        {rowData.map((row, index) => (
          <tr key={index}>
            {Object.keys(row).map((item, index) => (
              <Td key={index}>{row[item]}</Td>
            ))}
          </tr>
        ))}
      </tbody>
    </StyledTable>
  );
};

Table.defaultProps = {
  colDefs: [],
  rowData: []
};

export default Table;
