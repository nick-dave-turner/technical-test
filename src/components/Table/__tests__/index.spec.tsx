import React from "react";
import { shallow } from "enzyme";

import Table from "../";

const colDefs = ["Test Column 1", "Test Column 2"];

const rowData = [
  {
    col1: "Row 1 Column 1",
    col2: "Row 1 Column 2"
  },
  {
    col1: "Row 2 Column 1",
    col2: "Row 2 Column 2"
  }
];

describe("Table component", () => {
  describe("When required props are empty", () => {
    it("should match snapshot", () => {
      const component = shallow(<Table colDefs={[]} rowData={[]} />);
      expect(component).toMatchSnapshot();
    });
  });

  describe("When required props are complete", () => {
    it("should match snapshot", () => {
      const component = shallow(<Table colDefs={colDefs} rowData={rowData} />);
      expect(component).toMatchSnapshot();
    });
  });
});
