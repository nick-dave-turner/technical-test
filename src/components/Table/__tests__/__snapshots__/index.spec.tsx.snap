// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component When required props are complete should match snapshot 1`] = `
<styled.table>
  <thead>
    <tr>
      <Styled(styled.td)
        key="0"
      >
        Test Column 1
      </Styled(styled.td)>
      <Styled(styled.td)
        key="1"
      >
        Test Column 2
      </Styled(styled.td)>
    </tr>
  </thead>
  <tbody>
    <tr
      key="0"
    >
      <styled.td
        key="0"
      >
        Row 1 Column 1
      </styled.td>
      <styled.td
        key="1"
      >
        Row 1 Column 2
      </styled.td>
    </tr>
    <tr
      key="1"
    >
      <styled.td
        key="0"
      >
        Row 2 Column 1
      </styled.td>
      <styled.td
        key="1"
      >
        Row 2 Column 2
      </styled.td>
    </tr>
  </tbody>
</styled.table>
`;

exports[`Table component When required props are empty should match snapshot 1`] = `
<styled.table>
  <thead>
    <tr />
  </thead>
  <tbody>
    <tr>
      <styled.td
        colSpan={0}
      >
        <styled.p>
          You currently have no entries to display.
        </styled.p>
      </styled.td>
    </tr>
  </tbody>
</styled.table>
`;
